### PR TITLE
new subjson package + README updates

### PR DESCRIPTION
**Pull request by smyrman**
commit 69076cf5f3d0847665918788620130a7f4cbc2de (HEAD -> subjson, origin/subjson)
Author: Sindre Myren <sindre@searis.no>
Date:   Mon Dec 9 20:46:56 2019 +0100

    Add subjson package and update docs

    Adds a new package subjson containing middleware for parsing JSON byte
    values to Go types.

    Add an internal package to allow mocking of the `t *testing.T` parameter
    in example tests, as well as enforcing consistent output.

    Update README with a better introduction to both subtest and subjson.
    Correct spelling mistakes and perform minor adjustments in package
    documentation.

### CODE REVIEW

File: doc.go [hunk 1]
```diff
@@ -1,6 +1,6 @@
 // Package subtest provides a way of intializing small test functions suitable
 // for use as with the (*testing.T).Run method. Tests using package-defined
-// check functions can genrally be initalized in two main ways. here is an
+// check functions can genrally be initialized in two main ways. here is an
```
Comment (mmikalsen) [inline doc.go]: generally

File: examples/jsondiff/check_test.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+		err := pkg.SubsetOfJSON(a)(json.RawMessage(a))
+		t.Run("then the check should pass", subtest.Value(err).NoError())
+	})
+
+	t.Run("when b is checked against a", func(t *testing.T) {
+		err := pkg.SubsetOfJSON(a)(b)
+		t.Run("then the check should fail", subtest.Value(err).Error())
+	})
+
+	t.Run("when c is checked against a", func(t *testing.T) {
```
Comment (mmikalsen) [inline examples/jsondiff/check_test.go]: C ⊆ A  and C ⊊ A  is both true, C is a proper subset of A

File: examples/jsondiff/check_test.go [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+		// check if c is superset of a; normal syntax would be:
+		// t.Run(name, subtest.Value(c).Test(pkg.SupersetOfJSON(a)))
+		err := pkg.SupersetOfJSON(a)(c)
+		t.Run("then the check should fail", subtest.Value(err).Error())
+	})
+	t.Run("when a is cheked against c", func(t *testing.T) {
```
Comment (mmikalsen) [inline examples/jsondiff/check_test.go]: A is the SuperSet of C

File: examples/jsondiff/check_test.go (hunk 2 above)
Comment (smyrman) [reply examples/jsondiff/check_test.go]: There was a mistake in the test name. `"then the check should fail"` should have been `"then the check should pass"`

File: examples/jsondiff/check_test.go (hunk 3 above)
Comment (smyrman) [reply examples/jsondiff/check_test.go]: Improved the naming to be more clear. 

To be clear, we are using the following wording:
1. Given - the state of things before an action.
2. When - which action we are performing.
3. Then - what are the result(s) of the action.

Normally subtest functions wold be used for step 3, but when we are validating that they work, we instead use them  as part of step 2.

In other words, given `c` is a subset of `a`, we are testing _if_ `a` is decodable to JSON _and_ it is a JSON superset of `c`. After we have carried this out, the final `then` condition to compare the results.

**Pull request merged**